### ISSUE TITLE
Feed CNN features to LSTM

### ISSUE THREAD

**Issue by trminh89** [keras-team/keras#6426]
I want to build an end-to-end trainable model with the following proprieties:
- CNN to extract features from image 
- The features is reshaped to a matrix
- Each row of this matrix is then fed to LSTM1
- Each column of this matrix is then fed to LSTM2
- The output of LSTM1 and LSTM2 are concatenated for the final output

(it's more or less similar to Figure 2 in this paper: https://arxiv.org/pdf/1611.07890.pdf)

My problem now is after the reshape, how can I feed the values of feature matrix to LSTM with Keras?
This is my code so far with VGG16 net:
    
    # VGG16
    model = Sequential()
    model.add(Conv2D(64, (3, 3), activation='relu', padding='same', input_shape=(224, 224, 3)))
    model.add(Conv2D(64, (3, 3), activation='relu'))
    model.add(MaxPooling2D((2, 2)))
    
    # block 2
    model.add(Conv2D(128, (3, 3), activation='relu', padding='same'))
    model.add(Conv2D(128, (3, 3), activation='relu'))
    model.add(MaxPooling2D((2, 2)))
    
    # block 3
    model.add(Conv2D(256, (3, 3), activation='relu', padding='same'))
    model.add(Conv2D(256, (3, 3), activation='relu'))
    model.add(Conv2D(256, (3, 3), activation='relu'))
    model.add(MaxPooling2D((2, 2)))
    
    # block 4
    model.add(Conv2D(512, (3, 3), activation='relu', padding='same'))
    model.add(Conv2D(512, (3, 3), activation='relu'))
    model.add(Conv2D(512, (3, 3), activation='relu'))
    model.add(MaxPooling2D((2, 2)))
    
    # block 5
    model.add(Conv2D(512, (3, 3), activation='relu', padding='same'))
    model.add(Conv2D(512, (3, 3), activation='relu'))
    model.add(Conv2D(512, (3, 3), activation='relu'))
    model.add(MaxPooling2D((2, 2)))
    
    # block 6
    model.add(Flatten())
    model.add(Dense(4096, activation='relu'))
    model.add(Dense(4096, activation='relu'))
    
    # reshape the  feature 4096 = 64 * 64
    model.add(Reshape((64, 64)))
    
    # How to feed each row of this to LSTM?
    # This is my first solution but it doesn’t look correct: 
    # model.add(LSTM(256, input_shape=(64, 1)))  # 256 hidden units, sequence length = 64, feature dim = 1

**Comment by superMDguy** [comment]
You're going to have to use the [Functional API](https://keras.io/getting-started/functional-api-guide/).

**Comment by trminh89** [comment]
Could you please be more specific? @superMDguy 
@fchollet Any suggestions? Thanks!

**Comment by superMDguy** [comment]
First, make sure you import `Input` (`from keras.layers import Input`).  Next, try this:

```python
for i in range(64):
  input = Input(shape=(64,1))
  lstm = LSTM(256, input_shape=(64,1), name='row_'+str(i))(input)

**Comment by nqanh** [comment]
huh, interesting problem! I think the main problem is the end-to-end part. If you train 2 nets separately, then it's easy to reshape the feature vectors and feed them to LSTM. But how can we do it on the fly?

**Comment by trminh89** [comment]
This is my new code with Functioal API, but still I'm stuck at LSTM part :(


    def VGG16(weights_path=None): 
        model = Sequential()
        # block 1
        model.add(Conv2D(64, (3, 3), activation='relu', padding='same', input_shape=(224, 224, 3)))
        model.add(Conv2D(64, (3, 3), activation='relu'))
        model.add(MaxPooling2D((2, 2)))
        
        # block 2
        model.add(Conv2D(128, (3, 3), activation='relu', padding='same'))
        model.add(Conv2D(128, (3, 3), activation='relu'))
        model.add(MaxPooling2D((2, 2)))
        
        # block 3
        model.add(Conv2D(256, (3, 3), activation='relu', padding='same'))
        model.add(Conv2D(256, (3, 3), activation='relu'))
        model.add(Conv2D(256, (3, 3), activation='relu'))
        model.add(MaxPooling2D((2, 2)))
        
        # block 4
        model.add(Conv2D(512, (3, 3), activation='relu', padding='same'))
        model.add(Conv2D(512, (3, 3), activation='relu'))
        model.add(Conv2D(512, (3, 3), activation='relu'))
        model.add(MaxPooling2D((2, 2)))
        
        # block 5
        model.add(Conv2D(512, (3, 3), activation='relu', padding='same'))
        model.add(Conv2D(512, (3, 3), activation='relu'))
        model.add(Conv2D(512, (3, 3), activation='relu'))
        model.add(MaxPooling2D((2, 2)))
        
        # final block
        model.add(Flatten())
        model.add(Dense(4096, activation='relu'))
        model.add(Dense(4096, activation='relu'))
        
        # reshape 4096 = 64 * 64
        model.add(Reshape((64, 64)))
    
        if weights_path:
            model.load_weights(weights_path)
    
        return model
    
    
    if __name__ == '__main__':
        img_model = VGG16(None)
        img_input = Input(shape=(224, 224, 3))    
    
        encoded_img = img_model(img_input)  
        print 'encoded img shape: ', encoded_img.shape        
        
        # How to feed encoded_img to LSTM?

**Comment by andersonzhu** [comment]
@trminh89, I am having the same question. Have you figured it out?

**Comment by ahrnbom** [comment]
I think the Sequential model is fine, but you need TimeDistributed on all the non-LSTM layers so that the entire sequence of images are handled all the way up to the LSTM.
